pkg/controller/contrailcni/contrailcni_controller.go

```diff
@@ -0,0 +1,171 @@
+// and what is in the ContrailCNI.Spec
+func (r *ReconcileContrailCNI) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ContrailCNI")
+	instanceType := "contrailcni"
+	instance := &v1alpha1.ContrailCNI{}
+	ctx := context.TODO()
+
+	if err := r.Client.Get(ctx, request.NamespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if !instance.GetDeletionTimestamp().IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	if instance.Spec.ServiceConfiguration.ControlInstance != "" {
+		controlInstance := v1alpha1.Control{}
+		configInstance := v1alpha1.Config{}
+		configActive := configInstance.IsActive(instance.Labels["contrail_cluster"],
+			request.Namespace, r.Client)
+
+		controlActive := controlInstance.IsActive(instance.Spec.ServiceConfiguration.ControlInstance,
+			request.Namespace, r.Client)
+
+		if !configActive || !controlActive {
+			return reconcile.Result{}, nil
+		}
+	}
+
+	contrailCNIConfigName := request.Name + "-" + instanceType + "-configuration"
+	if err := r.configMap(contrailCNIConfigName, instanceType, instance).ensureContrailCNIConfigExists(r.ClusterInfo); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	cniDirs := CniDirs{
+		BinariesDirectory: r.ClusterInfo.CNIBinariesDirectory(),
+		DeploymentType:    r.ClusterInfo.DeploymentType(),
+	}
+
+	var nodesListOptions client.MatchingLabels = instance.Spec.CommonConfiguration.NodeSelector
+	var nodes corev1.NodeList
+	if err := r.Client.List(ctx, &nodes, nodesListOptions); err != nil && errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+	jobReplicas := int32(len(nodes.Items))
+	job := GetJob(cniDirs, request.Name, instanceType, &jobReplicas)
+	for idx, container := range job.Spec.Template.Spec.Containers {
+		instanceContainer := utils.GetContainerFromList(container.Name, instance.Spec.ServiceConfiguration.Containers)
+		if instanceContainer != nil {
+			(&job.Spec.Template.Spec.Containers[idx]).Image = instanceContainer.Image
+		}
+	}
+
+	if err := instance.PrepareJob(job, instance, request, r.Scheme, r.Client); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var clusterJob batchv1.Job
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: job.Name, Namespace: job.Namespace}, &clusterJob); err == nil {
+		if *clusterJob.Spec.Completions != jobReplicas || clusterJob.Labels["controller_generation"] != job.Labels["controller_generation"] {
+			_ = r.Client.Delete(ctx, &batchv1.Job{ObjectMeta: v1.ObjectMeta{Name: job.Name, Namespace: job.Namespace}}, client.PropagationPolicy("Background"))
+			return reconcile.Result{RequeueAfter: 5}, nil
+		}
+	} else if errors.IsNotFound(err) {
+		if err := r.Client.Create(ctx, job); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
```

Comment:
Don't we need an `else` here where we return error?

Reply:
Done